website/addons/box/model.py

```diff
@@ -0,0 +1,340 @@
+            node_settings.deauthorize(Auth(self.owner))
+            node_settings.save()
+        return self
+
+    def get_credentialsv2(self):
+        return CredentialsV2(
+            self.access_token, 
+            self.refresh_token, 
+            settings.BOX_KEY,
+            settings.BOX_SECRET,
+            self.token_refreshed_callback,
+        )
+
+    def __repr__(self):
+        return u'<BoxUserSettings(user={self.owner.username!r})>'.format(self=self)
+
+
+class BoxNodeSettings(AddonNodeSettingsBase):
+
+    user_settings = fields.ForeignField(
+        'boxusersettings', backref='authorized'
+    )
+
+    folder = fields.StringField(default=None)
+    folder_id = fields.IntegerField(default=None)
+
+    #: Information saved at the time of registration
+    #: Note: This is unused right now
+    registration_data = fields.DictionaryField()
+
+
+    @property
+    def display_name(self):
+        return '{0}: {1}'.format(self.config.full_name, self.folder)
+
+    @property
+    def has_auth(self):
+        """Whether an access token is associated with this node."""
+        return bool(self.user_settings and self.user_settings.has_auth and refresh_creds_if_necessary(self.user_settings))
+
+    def set_folder(self, folder, folder_id, auth):
+        self.folder = folder
+        self.folder_id = folder_id
+        # Add log to node
+        nodelogger = BoxNodeLogger(node=self.owner, auth=auth)
+        nodelogger.log(action="folder_selected", save=True)
+
+    def set_user_auth(self, user_settings):
+        """Import a user's Box authentication and create a NodeLog.
+
+        :param BoxUserSettings user_settings: The user settings to link.
+        """
+        self.user_settings = user_settings
+        nodelogger = BoxNodeLogger(node=self.owner, auth=Auth(user_settings.owner))
+        nodelogger.log(action="node_authorized", save=True)
+
+    def find_or_create_file_guid(self, path):
+        return BoxFile.get_or_create(self.owner, clean_path(os.path.join(self.folder, path.lstrip('/'))))
+
+    # TODO: Is this used? If not, remove this and perhaps remove the 'deleted' field
+    def delete(self, save=True):
+        self.deauthorize(add_log=False)
+        super(BoxNodeSettings, self).delete(save)
+
+    def deauthorize(self, auth=None, add_log=True):
+        """Remove user authorization from this node and log the event."""
+        node = self.owner
+        folder = self.folder
+
+        self.folder = None
+        self.folder_id = None
+        self.user_settings = None
+
+        if add_log:
+            extra = {'folder': folder}
+            nodelogger = BoxNodeLogger(node=node, auth=auth)
+            nodelogger.log(action="node_deauthorized", extra=extra, save=True)
+
+    def serialize_waterbutler_credentials(self):
+        if not self.has_auth:
+            raise exceptions.AddonError('Addon is not authorized')
+        return {'token': self.user_settings.access_token}
```

Comment:
It is being attempted here. Any time has_auth is called it will refresh the tokens if needed.